🚀 3단계 - 장바구니(담기)

ChaMinZi opened the pull request:
안녕하세요~ 2단계 질문 답변과 피드백 자세히 적어주셔서 감사합니다! 
3단계 미션 구현하여 리뷰 요청드립니다! 🙇‍♀️

## 구현사항
1. 장바구니 화면 구현
    - 제품 상세 화면에서 장바구니 담기 버튼 클릭 시 장바구니에 상품 담기
    - 장바구니 상품 개수 조정
        - 수량을 1보다 작게 하면 장바구니에서 상품을 제거
    - 장바구니 상품 삭제
    - 담긴 상품 가격의 총합이 주문하기 버튼에 표시

2. 장바구니 테스트 코드 추가

----------------------------------------------------------------
## Screen shot
https://github.com/user-attachments/assets/48ef0886-b012-4c39-a335-6273ff9dc46b

ChaMinZi committed:
chore: MergeConflict

ChaMinZi committed:
feat: 상품 상세 화면에서 장바구니 이동 추가

ChaMinZi committed:
Style: 배경 색상 변경

ChaMinZi committed:
docs: 3단계 구현 필요 사항 업데이트

ChaMinZi committed:
feat: 장바구니에 담긴 상품 UI 추가

ChaMinZi committed:
feat: Product.price type 변경

ChaMinZi committed:
feat: 하단 버튼 컴포넌트 분리

ChaMinZi committed:
feat: 상품 장바구니 담기 구현

ChaMinZi committed:
feat: 장바구니 상품 개수 조정, 상품 삭제 기능 추가

ChaMinZi committed:
refactor: resource 추가 및 Component 명 변경

ChaMinZi committed:
refactor: string Resource 적용

ChaMinZi committed:
feat: 장바구니 화면 테스트 코드 추가

lee-ji-hoon commented on app/src/androidTest/java/nextstep/shoppingcart/CartScreenTest.kt:
```diff
@@ -0,0 +1,110 @@
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    @Before
+    fun setup() {
+        composeTestRule.setContent {
+            ShoppingCartScreen(
+                onItemAdd = { product -> Cart.addOne(product) },
+                onItemRemove = { product -> Cart.removeOne(product) },
+                onDelete = { product -> Cart.removeAll(product) }
+            )
+        }
+    }
```
지금 전반적으로 Cart에 대한 의존성이 Test에도 깊게 나타나고 있는데요.

이렇게 됐을 때 2가지 정도의 문제점이 바로 보이네요.

```
1. Cart 라는 UI 외부에 대한 의존 (ViewModel 혹은 Repository 같은)
2. Object로 구현을 했기에 계속 clear해주는 동작이 필요
```

그러면 이것을 아래와 같이 Stateful 한 Screen과 Stateless한 Screen을 나누는 것입니다.

```kotlin
@Composable
fun ShoppingCartScreen(
    modifier: Modifier = Modifier,
    onBackClick: () -> Unit = {},
    onItemAdd: (Product) -> Unit = {},
    onItemRemove: (Product) -> Unit = {},
    onDelete: (Product) -> Unit = {},
) {
    val products: List<CartItem> = Cart.items
    val totalPrice by remember { derivedStateOf { Cart.totalPrice } }

    ShoppingCardScreen(
        onBackClick = onBackClick,
        totalPrice = totalPrice,
        products = products,
        onItemAdd = onItemAdd,
        onItemRemove = onItemRemove,
        onDelete = onDelete,
        modifier = modifier
    )
}

@Composable
internal fun ShoppingCardScreen(
    onBackClick: () -> Unit,
    totalPrice: Int,
    products: List<CartItem>,
    onItemAdd: (Product) -> Unit,
    onItemRemove: (Product) -> Unit,
    onDelete: (Product) -> Unit,
    modifier: Modifier = Modifier,
) { ... }
```

이 형태로 하고 Stateless 한 ShoppingCardScreen 을 테스트를 해보시면 Cart에 대한 의존 없이 순수한 UI 테스트가 가능해집니다

lee-ji-hoon commented on app/src/main/java/nextstep/shoppingcart/ui/shoppingcart/ShoppingCartScreen.kt:
```diff
@@ -1,23 +1,63 @@
+    onDelete: (Product) -> Unit = {},
 ) {
+    val products = Cart.items
+    val totalPrice by remember { derivedStateOf { Cart.totalPrice } }
```
`totalPrice` 에서 `derivedStateOf` 에 대한 이점을 있었을까요? `derivedStateOf` 가 무엇이고 어떠한 상황에서 사용하는지 확인해보면 좋을 것 같네요!

ChaMinZi replied:
넵넵, derivedStateOf가 변경이 적게 일어나니까 막연히 사용했던것 같아요!

입력에 비해 출력 값이 변화가 적어서 불필요한 리컴포지션을 막을 때 사용 이점이 있는데,
위의 UI에서는 상품이 추가/삭제되면서 항상 totalPirce가 변경이 일어나기에 딱히 이점이 없어 보여요.

항상 신경써서 코드를 작성해야 하는데 그렇지 못했네요🥲🥲
리뷰 감사합니다👍

lee-ji-hoon commented on app/src/main/java/nextstep/shoppingcart/ui/shoppingcart/ShoppingCartScreen.kt:
```diff
@@ -1,23 +1,63 @@
             ShoppingCartTopAppBar(onBackClick = onBackClick)
-        }
+        },
+        bottomBar = {
```
Scaffold의 bottomBar 사용 좋습니다 👍

lee-ji-hoon commented on app/src/main/java/nextstep/shoppingcart/ui/shoppinglist/ShoppingItemView.kt:
```diff
@@ -30,12 +31,14 @@ fun ShoppingItem(
             .clickable(onClick = { onItemClick(product) }),
     ) {
         AsyncImage(
-            modifier = modifier.fillMaxWidth(),
+            modifier = modifier
+                .fillMaxWidth()
+                .aspectRatio(1f),
```
제가 좀 뒤늦게 확인한 것인데, modifier는 최상위 Composable에 할당을 하는 것이 좋습니다.

```kotlin
@Composable
fun SampleScreen() {
    ShoppingItemView(
        modifier = Modifier.background(Color.Red)
    )
}
```

위 같은 상황에서 당연히 최상위 Modifier에 할당이 될 것이라고 생각을 하게 되지만 현재 구현은 AsyncImage에 걸려있기에 하위 Text들 영역까지 Background가 안되는 문제가 생기기도 합니다.

lee-ji-hoon reviewed (changes requested):
안녕하세요 민지님 3단계도 잘 구현을 해주셨습니다.

단 테스트 관점에서 Stateful Stateless 컴포넌트를 나눈 것은 어떨지 리뷰를 남겼으며 지금 확인하고 테스트 코드를 수정을 해보면 좋을 것 같네요.

이번 단계를 지나면 4단계니 마지막까지 화이팅입니다!

ChaMinZi committed:
docs: 3단계 리뷰 사항 업데이트

ChaMinZi committed:
refactor: 3단계 리뷰 사항 업데이트

change
- 불필요한 derivedStateOf 제거
- ShoppingCartScreen을 Stateless한 Screen과 Stateful한 screen으로 나누기

ChaMinZi committed:
refactor: 공통된 Modifier 요소 최상위에서 할당하도록 변경

ChaMinZi committed:
bugfix: 특정 기기에서 색상 다르게 노출되는 이슈 수정

ChaMinZi committed:
refactor: totalPrice를 observable 하게 변경

ChaMinZi committed:
refactor: TestCode에서 Cart에 대한 의존성 제거

ChaMinZi committed:
refactor: TestCode totalPrice 변경 시, UI update되도록 변경

ChaMinZi commented:
안녕하세요! 3단계 리뷰 사항들 반영하였습니다!
확인 부탁드립니다~

- [x] 불필요한 derivedStateOf 제거
    - totalPrice 를 observable 변수로 변경

- [x]  ShoppingCartScreen을 Stateless한 Screen과 Stateful한 screen으로 나누기
    - ShoppringCartScreen UI Test 코드에서 Cart class에 대한 의존성 제거

- [x] 공통된 Modifier 요소 최상위에서 할당하도록 변경

-----------
리뷰 사항을 반영하다 보니 궁금한 점이 생겼습니다!

Cart에 대한 의존 없이 순수한 UI 테스트를 위해 stateless한 화면을 만드는 것은 이해했습니다~
하지만 `ClickListener`(`onItemAdd`/`onItemRemove`)에 `Cart` 클래스 내부 구현과 거의 동일한 코드를 작성하게 되어 이 부분이 조금 걸립니다. UI 테스트만을 위해서는 ViewModel이나 repository(`위의 코드에서는 Cart`)에 의존하지 않는 것이 좋다는 점은 알겠지만, 유지보수성 측면에서 거의 동일한 코드를 두 곳에 작성하는 것이 맞는지 고민이 됩니다.

이전 회원가입 미션과 같이 내부 구현만 다른 클래스로 빼서 파라미터와 리턴값으로 로직 부분을 분리하는 방법도 있을 것 같습니다. 
그러나 한편으로는 하나의 기능 개발에 이 정도로 클래스를 분리하는 것이 적절한지에 대해서도 고민이 됩니다.
멘토님께서는 테스트를 위해 어느 정도까지 클래스를 분리하고 적용하시는지 의견을 여쭙고 싶습니다.

감사합니다!!

lee-ji-hoon commented on app/src/androidTest/java/nextstep/shoppingcart/CartScreenTest.kt:
```diff
@@ -0,0 +1,140 @@
+            ShoppingCartScreen(
+                products = cartItems,
+                totalPrice = totalPrice,
+                onItemAdd = { product ->
+                    val item = cartItems.find { it.product == product }
+                    if (item == null) {
+                        cartItems.add(CartItem(product, 1))
+                    } else {
+                        val index = cartItems.indexOf(item)
+                        cartItems[index] = item.copy(count = item.count + 1)
+                    }
+                    updateTotalPrice()
+                },
```
> Cart에 대한 의존 없이 순수한 UI 테스트를 위해 stateless한 화면을 만드는 것은 이해했습니다~
하지만 ClickListener(onItemAdd/onItemRemove)에 Cart 클래스 내부 구현과 거의 동일한 코드를 작성하게 되어 이 부분이 조금 걸립니다. UI 테스트만을 위해서는 ViewModel이나 repository(위의 코드에서는 Cart)에 의존하지 않는 것이 좋다는 점은 알겠지만, 유지보수성 측면에서 거의 동일한 코드를 두 곳에 작성하는 것이 맞는지 고민이 됩니다.

`UI Test` 의 관점으로만 봤을 때 `lambda` 로 어떤 것이 실행이 되든 전 크게 신경을 안 쓴다. 정확히는 `lambda` 로 지금처럼 비즈니스적인 로직을 테스트를 굳이 해야 할까?라는 관점입니다.

```
어떠한 State가 들어왔을 때 어떻게 표시가 되는가.
```

이게 메인이 되어야 하지 않았을까 싶습니다.

```
1. + 버튼을 누르면 아이템이 증가한다.
2. 아이템이 증가했으니 UI가 변경이 된다.
```

여기서 1번의 명제를 지키기 위해서 `lambda`에서 이것저것을 하다 보니 이렇게 된 것인데, 해당 로직은 비즈니스 로직이기에 굳이 `UI Test` 에서는 1번의 내용을 검증을 하기보다는 `어떠한 상태가 들어왔을 때 UI가 제대로 표시가 되는지` 를 우선적으로 테스트 해보는 것은 어떨까요?

lee-ji-hoon reviewed (approved):
민지님 안녕하세요! 리뷰 반영 해주신 부분 잘 확인했으며, 궁금한 점에 대해서는 제가 추가로 리뷰를 남겨놨습니다.

해당 부분에 대해서는 마지막 4단계를 진행하면서 고민을 해봐도 충분할 것으로 보여서 다음 단계를 진행하면서 고민을 해보시면 좋을 것 같네요!

마지막까지 포기하지 않고 화이팅입니다!
